### PR TITLE
[graphqlErrorHandler] Include upstream HTTP status code.

### PR DESCRIPTION
**Pull request by alloy**
Part of https://artsyproduct.atlassian.net/browse/DISCO-671

Augments an error with `extensions`, as described in the spec https://facebook.github.io/graphql/draft/#sec-Errors.

screenshot 2019-01-11 at 14 09 34

### CODE REVIEW

File: src/lib/graphqlErrorHandler.ts [hunk 1]
```diff
@@ -63,46 +74,56 @@ const reportErrorToSentry = (
...
-            reportErrorToSentry(error, queryContext)
-          }
+      flattenedErrors.forEach(e => {
+        if (shouldReportParentError(e.originalError)) {
+          reportErrorToSentry(e, queryContext)
         }
-      } else {
-        reportErrorToSentry(error, queryContext)
-      }
+      })
```
Comment (alloy) [inline src/lib/graphqlErrorHandler.ts]: @dleve123 Refactored this a bit after all.

File: src/lib/graphqlErrorHandler.ts [hunk 2]
```diff
@@ -63,46 +74,56 @@ const reportErrorToSentry = (
   })
 }
 
+type WriteablePartial<T> = { -readonly [P in keyof T]+?: T[P] }
```
Comment (alloy) [inline src/lib/graphqlErrorHandler.ts]: I know, right?

There's a bit more documentation [here](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#improved-control-over-mapped-type-modifiers). In short, we're mapping over all the properties in the given object structure and removing `readonly` modifiers with the `-` sign and adding the optional (`?`) modifier with the `+` sign.

Comment (mzikherman) [reply src/lib/graphqlErrorHandler.ts]: Interesting!

I'm understanding why this is the desired typing for `result`, and also marveling at the actual type declaration!

Comment (oxaudo) [reply src/lib/graphqlErrorHandler.ts]: This is just crazy! This line:)

Comment (alloy) [reply src/lib/graphqlErrorHandler.ts]: Yeah it’s definitely not the easiest 😄, but I trust we're at the point that we can introduce some of these harder ones when _really_ useful and everybody still being able to follow along, given some notes. (And to be clear, this applies to me too, this comment is as much for others as it is for me.)

**Pull request merged**